presc/copies/sampling.py

```diff
@@ -40,6 +41,118 @@ def dynamical_range(df):
     return range_dict
 
 
+def find_categories(df, add_NaNs=False):
+    """Returns the categories of the dataset features.
+
+    Parameters
+    ----------
+    df : pandas DataFrame
+        The dataset with all the categorical features to analyze.
+    add_NaNs : bool
+        If True the sampler adds a "NaNs" category for the features that have
+        any null values and assigns it the appropriate fraction.
+
+    Returns
+    -------
+    dict of dicts
+        A dictionary with an entry per dataset feature (dictionary keys are the
+        column names), where each feature entry contains a nested dictionary
+        with its categories and the fraction of each category present in the
+        analyzed dataset (the nested dictionary key for this information is
+        "categories", which is also a dictionary with one entry per category).
+    """
+    categories_dict = {}
+    for feature in df:
+        if is_discrete(df[feature]):
+            # Remove NaN values from selection
+            df_no_nans = df[df[feature].notnull()]
```

Comment:
I see you're using `value_counts` inside the for loop below. In fact, all of this computation could be handled by a single call to `value_counts`. Eg.
```python
value_props = df[feature].value_counts(normalize=True, dropna=(not add_nans))
```
Then `value_props` is Series with an entry for each unique value listing its relative frequency, and you can do
```python
categories_dict[feature] = {
    "categories": value_props.to_dict()
}
```
It'll handle the cases around NaNs implicitly using `dropna`.

Reply:
I leave it as it is for the moment, as we agreed in our discussion.